File: _ref/reader-dev/web-ui/src/api/rss.ts
import { get, post, type RequestOptions } from './request'
import type { RssArticle, RssSource, ReturnData } from '@/types'

/** GET /reader3/getRssSources：当前用户 RSS 订阅源列表 */
export function getRssSources(): Promise<ReturnData<RssSource[]>> {
  return get<RssSource[]>('/getRssSources')
}

/** POST /reader3/saveRssSource：新增/更新订阅源（body = 完整订阅源 JSON） */
export function saveRssSource(source: RssSource): Promise<ReturnData<null>> {
  return post<null>('/saveRssSource', source)
}

/** POST /reader3/deleteRssSource：删除订阅源（body { rssSourceUrl }） */
export function deleteRssSource(rssSourceUrl: string): Promise<ReturnData<null>> {
  return post<null>('/deleteRssSource', { rssSourceUrl })
}

/**
 * GET /reader3/getRssArticles：订阅源文章列表（params rssSourceUrl + page，分页；hasRead 已读标记）。
 * sortUrl 为 legacy sortUrl 多段 `名称::地址` 中的分类 URL——传该段后后端抓对应分类 feed。
 * 后端每次调用会重新抓取 feed——「刷新全部」即逐源循环调此接口（silent 模式不弹全局提示）。
 */
export function getRssArticles(
  rssSourceUrl: string,
  page = 1,
  sortUrl?: string,
  opts?: RequestOptions,
): Promise<ReturnData<RssArticle[]>> {
  return get<RssArticle[]>('/getRssArticles', { rssSourceUrl, page, sortUrl }, opts)
}

/** POST /reader3/markRssArticleRead：标记文章已读/未读（body { articleUrl, read }） */
export function markRssArticleRead(articleUrl: string, read: boolean): Promise<ReturnData<null>> {
  return post<null>('/markRssArticleRead', { articleUrl, read })
}

/** GET /reader3/getRssArticle：文章正文（data: { content }，content 为 HTML） */
export function getRssArticle(url: string): Promise<ReturnData<{ content: string }>> {
  return get<{ content: string }>('/getRssArticle', { url })
}

File: _ref/reader-dev/src/model/rss.rs
//! RSS 实体（兼容 legacy RssSource / RssArticle 字段子集）
//!
//! - RssSource：表列仅保留任务规格字段（rss_source_url / rss_source_name / rss_source_group /
//!   enabled / user_namespace / raw_json），完整 legacy JSON 原文保存在 raw_json（保底不丢字段）；
//!   API 返回时以 raw_json 为基底、表列为覆盖（见 router::rss_source_json）
//! - RssArticle：url 为主键，content 为 feed 正文/摘要（HTML 或文本）

use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// RSS 源（表：rss_sources）
#[derive(Debug, Clone, Default, Serialize, Deserialize, FromRow)]
#[serde(default)]
pub struct RssSource {
    #[serde(rename = "sourceUrl")]
    #[sqlx(rename = "rss_source_url")]
    pub source_url: String,
    #[serde(rename = "sourceName")]
    #[sqlx(rename = "rss_source_name")]
    pub source_name: String,
    #[serde(rename = "sourceGroup")]
    #[sqlx(rename = "rss_source_group")]
    pub source_group: Option<String>,
    pub enabled: bool,
    #[serde(skip)]
    #[sqlx(rename = "user_namespace")]
    pub user_namespace: String,
    #[serde(skip)]
    #[sqlx(rename = "raw_json")]
    pub raw_json: Option<String>,
}

impl RssSource {
    /// 从 raw_json（完整 legacy JSON 原文）读取表列之外的字段
    fn json_field(&self, key: &str) -> Option<String> {
        self.raw_json
            .as_deref()
            .and_then(|r| serde_json::from_str::<serde_json::Value>(r).ok())
            .and_then(|v| v.get(key).and_then(|x| x.as_str()).map(str::to_string))
    }

    /// 请求头（legacy header 字段）
    pub fn header(&self) -> Option<String> {
        self.json_field("header")
    }

    /// 排序/列表 URL（legacy sortUrl 字段）
    pub fn sort_url(&self) -> Option<String> {
        self.json_field("sortUrl")
    }

    /// 列表规则（legacy ruleArticles）
    pub fn rule_articles(&self) -> Option<String> {
        self.json_field("ruleArticles")
    }

    /// 下一页规则（legacy ruleNextPage；"PAGE" = 同一 URL 按页码继续）
    pub fn rule_next_page(&self) -> Option<String> {
        self.json_field("ruleNextPage")
    }

    /// 标题规则（legacy ruleTitle）
    pub fn rule_title(&self) -> Option<String> {
        self.json_field("ruleTitle")
    }

    /// 发布时间规则（legacy rulePubDate）
    pub fn rule_pub_date(&self) -> Option<String> {
        self.json_field("rulePubDate")
    }

    /// 摘要/描述规则（legacy ruleDescription）
    pub fn rule_description(&self) -> Option<String> {
        self.json_field("ruleDescription")
    }

    /// 配图规则（legacy ruleImage）
    pub fn rule_image(&self) -> Option<String> {
        self.json_field("ruleImage")
    }

    /// 文章链接规则（legacy ruleLink）
    pub fn rule_link(&self) -> Option<String> {
        self.json_field("ruleLink")
    }

    /// 文章正文规则（legacy ruleContent）
    pub fn rule_content(&self) -> Option<String> {
        self.json_field("ruleContent")
    }

    /// 订阅源图标（legacy sourceIcon）
    pub fn source_icon(&self) -> Option<String> {
        self.json_field("sourceIcon")
    }

    /// 并发率（legacy concurrentRate）
    pub fn concurrent_rate(&self) -> Option<String> {
        self.json_field("concurrentRate")
    }
}

/// RSS 文章（表：rss_articles）
#[derive(Debug, Clone, Default, Serialize, Deserialize, FromRow)]
#[serde(default, rename_all = "camelCase")]
pub struct RssArticle {
    /// 文章链接（主键）
    pub url: String,
    /// 所属 RSS 源 URL
    #[sqlx(rename = "source_url")]
    pub source_url: String,
    pub title: String,
    pub author: String,
    /// 发布时间（毫秒时间戳）
    pub time: i64,
    /// 正文（feed content/summary，或抓取网页提取）
    pub content: Option<String>,
    /// 封面/配图
    pub cover: Option<String>,
    /// 已读标记（JSON 输出为 hasRead；SQLite 列 read）
    #[serde(rename = "hasRead")]
    pub read: bool,
    #[serde(skip)]
    #[sqlx(rename = "user_namespace")]
    pub user_namespace: String,
}
